src/tbb/misc_ex.cpp

```diff
@@ -335,38 +348,29 @@ int NumberOfProcessorGroups() {
     return ProcessorGroupInfo::NumGroups;
 }
 
-// Offset for the slot reserved for the first external thread
-#define HoleAdjusted(procIdx, grpIdx) (procIdx + (holeIdx <= grpIdx))
-
 int FindProcessorGroupIndex ( int procIdx ) {
-    // In case of oversubscription spread extra workers in a round robin manner
-    int holeIdx;
-    const int numProcs = theProcessorGroups[ProcessorGroupInfo::NumGroups - 1].numProcsRunningTotal;
-    if ( procIdx >= numProcs - 1 ) {
-        holeIdx = INT_MAX;
-        procIdx = (procIdx - numProcs + 1) % numProcs;
+    int current_grp_idx = ProcessorGroupInfo::HoleIndex;
+    if (procIdx > theProcessorGroups[ProcessorGroupInfo::HoleIndex].numProcs-1  && procIdx <= theProcessorGroups[ProcessorGroupInfo::NumGroups - 1].numProcsRunningTotal-1 ) {
+        procIdx = procIdx - theProcessorGroups[current_grp_idx].numProcs;
+        do {
+            current_grp_idx = (current_grp_idx + 1) % (ProcessorGroupInfo::NumGroups);
+            procIdx = procIdx - theProcessorGroups[current_grp_idx].numProcs;
+
+        } while (procIdx >= 0);
     }
-    else
-        holeIdx = ProcessorGroupInfo::HoleIndex;
-    __TBB_ASSERT( hardware_concurrency_info == do_once_state::initialized, "FindProcessorGroupIndex is used before AvailableHwConcurrency" );
-    // Approximate the likely group index assuming all groups are of the same size
-    int i = procIdx / theProcessorGroups[0].numProcs;
-    // Make sure the approximation is a valid group index
-    if (i >= ProcessorGroupInfo::NumGroups) i = ProcessorGroupInfo::NumGroups-1;
-    // Now adjust the approximation up or down
-    if ( theProcessorGroups[i].numProcsRunningTotal > HoleAdjusted(procIdx, i) ) {
-        while ( theProcessorGroups[i].numProcsRunningTotal - theProcessorGroups[i].numProcs > HoleAdjusted(procIdx, i) ) {
-            __TBB_ASSERT( i > 0, nullptr);
-            --i;
+    else if (procIdx > theProcessorGroups[ProcessorGroupInfo::NumGroups - 1].numProcsRunningTotal) {
```

Comment:
\> numProcsRunningTotal ***- 1***, right?

Reply:
Changed the logic a bit to accommodate this.